packages/create-pantheon-decoupled-kit/src/utils/dedupeTemplates.ts

```diff
@@ -0,0 +1,91 @@
+import chalk from 'chalk';
+import path from 'path';
+import klaw from 'klaw';
+import type { TemplateData, MergedPaths } from '../types';
+
+const rootDir = new URL('.', import.meta.url).pathname;
+
+interface Template {
+	addon: boolean;
+	template: string;
+	base: string;
+}
+
+export const getAllTemplates = async (
+	templateDir: string,
+	addon: boolean,
+): Promise<Template[]> => {
+	const templates: Template[] = [];
+	const base = path.basename(templateDir);
+	const fullTemplateDirPath = path.resolve(rootDir, 'templates', templateDir);
+
+	// loop through each template dir and gather all of the templates
+	// include the base and if it is an addon on the object
+	for await (const file of klaw(fullTemplateDirPath)) {
+		if (file.stats.isDirectory()) continue;
+		const tempObj: { addon: boolean; template: string; base: string } = {
+			addon,
+			base,
+			template: '',
+		};
+		const templateName = file.path.split(templateDir)[1];
+		tempObj.template = templateName;
+
+		templates.push(tempObj);
+	}
+	return templates;
+};
+
+/**
+ * @param unmergedTemplates an array of objects. Each object includes
+ * the template name, its base directory, and a boolean indicating whether
+ * this template is an addon or not
+ * @returns an object with keys equal to the paths merged. The value of
+ * each key includes the baseDir of the template, and the addon status
+ */
+export const mergeTemplatePaths = (unmergedTemplates: Template[]) => {
+	const mergedPaths: MergedPaths = {};
+
+	for (const { addon, template, base } of unmergedTemplates) {
+		if (!mergedPaths[template]) {
+			// set the template as the key and if it is an addon as the value
+			mergedPaths[template] = { addon, base };
+			// if we find the template already added to mergedPaths
+			// and it is an addon
+		} else if (mergedPaths[template] && !mergedPaths[template].addon) {
+			// if the current template is an addon, set the base to that template
+			delete mergedPaths[template];
+			mergedPaths[template] = { base, addon };
+		}
+	}
+	return mergedPaths;
+};
+
+/**
+ * In case there is a template path that exists in two or more generators,
+ * we want to favor the addon. This way, we avoid writing to a
```

Comment:
I'd agree that favoring addons makes sense, but I do have some questions about the details. Some of this is probably me not being fully in the loop on how addons work.

* If a base generator and and addon generator both operate on recipes.js, does only the addon template end up being added in this case? If that is true, I'm not sure it is 'bad', but it does change some of my assumptions about how addons would have to be structured.
* What happens if there are multiple addons that operate on the same file? Does the last addon win?

Reply:
> If a base generator and and addon generator both operate on recipes.js, does only the addon template end up being added in this case? 

Yes that is the case here. It's possible that we won't need to do this if we structure add-ons differently, but it seemed like the path of least resistance for now

> If that is true, I'm not sure it is 'bad', but it does change some of my assumptions about how addons would have to be structured.

I'm open to restructuring this if there's another pattern that makes more sense. I'm hoping that moving more things to packages and using more partials will allow us to avoid clashing template files in the first place. I think there's been a bit of a learning curve to use partials so we aren't using them as much as I would like at the moment but they may help solve some things.

> What happens if there are multiple addons that operate on the same file? Does the last addon win?
currently that is the case. We could use a number instead of a boolean for the `addon` property indicating it's priority (lower wins) or something like that in case we want more fine grained control over that, but for now the last called generator wins.

Reply:
Gotcha. I could imagine cases where this behavior could become a problem, but I'm just making stuff up :) Knowing the bounds of this helps, and if this behavior becomes an issue in the future, we can tackle it with a real-world use case.